File: 01_Arrays_Hashing/03_Two_Sum/main.py
def is_palindrome(s: str) -> bool:
    # Normalize the string: convert to lowercase and filter out non-alphanumeric characters
    filtered_chars = [c.lower() for c in s if c.isalnum()]
    
    # Use two pointers
    left, right = 0, len(filtered_chars) - 1
    
    # Compare characters from both ends
    while left < right:
        if filtered_chars[left] != filtered_chars[right]:
            return False
        left += 1
        right -= 1
    
    return True

# Test cases
s1 = "Was it a car or a cat I saw?"
s2 = "tab a cat"

print(f"Example 1: {is_palindrome(s1)}")  # Output: True
print(f"Example 2: {is_palindrome(s2)}")  # Output: False

File: 01_Arrays_Hashing/03_Two_Sum/main.c
#include <stdio.h>
#include <ctype.h> // for isalnum
#include <string.h> // for strlen

// Function to check if a string is a palindrome
int isPalindrome(char *s) {
    // Initialize pointers
    char *left = s;
    char *right = s + strlen(s) - 1;
    
    // Traverse the string
    while (left < right) {
        // Skip non-alphanumeric characters
        while (!isalnum(*left) && left < right) {
            left++;
        }
        while (!isalnum(*right) && left < right) {
            right--;
        }
        
        // Compare characters (case insensitive)
        if (tolower(*left) != tolower(*right)) {
            return 0; // Not a palindrome
        }
        
        // Move pointers inward
        left++;
        right--;
    }
    
    return 1; // Palindrome
}

int main() {
    // Test cases
    char s1[] = "Was it a car or a cat I saw?";
    char s2[] = "tab a cat";
    
    // Test and output results
    printf("Example 1: %s\n", isPalindrome(s1) ? "true" : "false");
    printf("Example 2: %s\n", isPalindrome(s2) ? "true" : "false");
    
    return 0;
}
